shelley-ma/formal-spec/value.tex

```diff
@@ -1,77 +1,225 @@
+When composed as
+
+\[\fun{coin} \circ \fun{inject} = \fun{id}_{\Coin}\]
+
+these functions give the identity on $\Coin$, but composed in the opposite order as
+
+\[\fun{inject} \circ \fun{coin}\]
+
+they return a $\ValClass$ containing only the Ada tokens contained in the original $\ValClass$.
+
+\textbf{Fungible and non-fungible tokens.}
+Recall that a token is said to be fungible at a specific time (slot $s$, for example) if
+there are two or more tokens in existence on the ledger that are indistinguishable from
+each other, ie. they have the exact same identifier. Tokens with the same
+identifier are fungible with each other, but not with tokens with a different identifier.
+
+If there exists only a single
+token with a given identifier, it is said to be unique (or non-fungible).
+The usual meaning of \emph{non-fungible token} implies, additionally to the
+requirement that a token is unique on the ledger at a given time, that
+it is programmatically guaranteed to remain unique indefinitely. This
+type of guarantee can be encoded in a minting policy.
+
+
+\subsection{Multi-assets and $\Value$ in the Mary Era}
+
+$\Value$ is the type to which
+$\ValClass$ is instantiated in the Mary era to support transacting with
+multiple types of assets, both user-defined and minted, and Ada. Terms of this
+type represent heterogeneous collections of assets. This
+This section describes the type $\Value$, the operations required on
+it, and its relation to $\Coin$.
+
+Figure \ref{fig:defs:value} introduces the basic types that make up the $\Value$ type,
+as well as some of the basic functions needed for $\ValClass$ instantiation.
+
+\begin{figure*}[t!]
   \emph{Derived types}
   %
   \begin{equation*}
     \begin{array}{r@{~\in~}l@{\qquad=\qquad}lr}
+      \var{aname} & ~\AssetName & \mathsf{ByteString} \\
       \var{pid} & \PolicyID & \ScriptHash \\
+      \var{aid} & \AssetID & \PolicyID \times \AssetName \\
       \var{quan} & \Quantity & \Z \\
       %\text{quantity of a token}\\
       \var{v}, \var{w} & \Value
-      & \PolicyID \to_0 ( \AssetName \to_0 \Quantity )
+      & \Coin \times \PolicyID \to_0 ( \AssetName \to_0 \Quantity )
 %      & \text{a collection of tokens}
     \end{array}
   \end{equation*}
   %
-  \emph{Abstract Functions and Values}
-  %
-  \begin{align*}
-    \mathsf{adaPolicy} \in& ~\PolicyID & \text{Ada Policy ID} \\
-    \mathsf{adaName} \in& ~\AssetName & \text{Ada Asset Name} \\
-    \mathsf{size} \in& ~\Value\to\MemoryEstimate & \text{the size of a Value}
-  \end{align*}
-  \emph{Helper Functions}
+  \emph{Functions required for $\ValClass$ instantiation with $\Value$}
   %
   \begin{align*}
-    & \fun{inject} \in \Coin\to \Value \\
-    & \fun{inject}~c = \lambda~\var{pid}~\var{aname}.
-      \begin{cases}
-        c & \var{pid} = \mathsf{adaPolicy}~\text{and}~\var{aname} = \mathsf{adaName} \\
-        0 & \text{otherwise}
-      \end{cases}
+    & \fun{pids} ~\in~ \Value \to \powerset{\PolicyID} \\
```

Comment:
This function should be part of the definition of ValClass

Reply:
Also, because `Coin` is now separated out in `Value`, `supp` isn't well-defined anymore. I think for the spec we should keep the old definition of `Value`, and mention that we use something else in the implementation.

Reply:
the reason `pids` (renamed to `policies` to match the code) and `anames` is not in `ValClass` is that it's not defined for Coin, nor does it need to be. These functions are not used anywhere on Coin.

Reply:
Oh, I see, `minted` is the function we use.